.github/container/jax_nsys/python/jax_nsys/jax_nsys/analysis.py

```diff
@@ -1,10 +1,135 @@
+        return 8
+
+    # S32 is a 32-bit type and so on.
+    # FIXME: does not handle FP8 yet
+    for prefix in ["BF", "C", "F", "S", "U"]:
+        if enum_name.startswith(prefix):
+            return int(enum_name[len(prefix) :])
+    raise Exception(f"Could not deduce size of {enum_name}")
+
+
+@functools.lru_cache
+def get_message_size(program_id: int, instruction: str) -> int:
+    """
+    Given the name of a collective instruction (e.g. all-gather-start.N), calculate the
+    message size in bytes. See https://openxla.org/xla/operation_semantics#allgather,
+    https://openxla.org/xla/operation_semantics#allreduce and so on for more explanation
+    of the semantics. This implementation aims to follow the same conventions that NCCL
+    uses in its NVTX payloads and tests.
+    """
+    module_proto = xla_module_metadata(program_id)
+    _, inst = module_proto.find_instruction(instruction)
+    assert inst.opcode in {
+        "all-gather-start",
+        "all-reduce-start",
+        "collective-broadcast",
+        "collective-permute-start",
+        "reduce-scatter",
+    }, f"{instruction}: message size calculation for {inst.opcode} has not yet been validated"
+    if inst.opcode == "collective-permute-start":
+        # See https://openxla.org/xla/operation_semantics#collectivepermute, which
+        # generates pair-wise send+recv between devices
+        collective_size = 2
+    else:
+        # replica_groups is something like {{0,1},{4,5},{2,3},{6,7}}, if there are 8
+        # devices that are doing pair-wise collectives
+        collective_sizes = tuple(
+            {len(group.replica_ids) for group in inst.replica_groups}
+        )
+        assert (
+            len(collective_sizes) == 1
+        ), f"Heterogeneous collective {inst.replica_groups} could not be interpreted"
+        collective_size = collective_sizes[0]
+    total_msg_size = 0
+    for operand_id in inst.operand_ids:
+        _, operand = module_proto.find_instruction_by_id(operand_id)
+        msg_size_bits = math.prod(
+            operand.shape.dimensions,
+            start=element_type_in_bits(operand.shape.element_type),
+        )
+        if inst.opcode == "reduce-scatter":
+            # NCCL's convention is that the message size of a reduce-scatter is the size of output buffer:
+            # https://github.com/NVIDIA/nccl/blob/ab2b89c4c339bd7f816fbc114a4b05d386b66290/src/collectives.cc#L122
+            assert msg_size_bits % collective_size == 0
+            msg_size_bits //= collective_size
+        assert msg_size_bits % 8 == 0
+        total_msg_size += msg_size_bits // 8
+
+    # Calculate the correction factor from algorithm bandwidth to bus bandwidth, see:
+    # https://github.com/NVIDIA/nccl-tests/blob/master/doc/PERFORMANCE.md#bus-bandwidth
+    collective = inst.opcode.removesuffix("-start")
+    bw_correction, bus_correction = {
```

Comment:
Not a fan of computing all cases to use a single one :-D I'd usually probably factor this into a helper. Fwiw, from Python 3.10 you get `match`, so this could become:
```
def collective_correction(kind: str, size: int) -> tuple[float, float]:
  match kind:
    case "all-gather": return (size, (size - 1) / size)
    case ...
    case _: assert False, f"unknown collective kind {kind}"
```